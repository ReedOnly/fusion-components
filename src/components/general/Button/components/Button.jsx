import React, { forwardRef } from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

const ButtonComponent = forwardRef(
    (
        {
            children,
            disabled,
            onMouseDown,
            onMouseUp,
            onClick,
            onClickCapture,
            ...props
        },
        ref
    ) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.displayName =
    "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
