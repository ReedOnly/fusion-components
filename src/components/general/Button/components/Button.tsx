import * as React from "react";
import Content from "./Content";
import getButtonClasses from "../buttonClasses";

type ButtonProps = {
    disabled?: boolean,
    onClick?: React.EventHandler<React.SyntheticEvent>,
    onClickCapture?: React.EventHandler<React.SyntheticEvent>,
    onMouseDown?: React.EventHandler<React.SyntheticEvent>,
    onMouseUp?: React.EventHandler<React.SyntheticEvent>,
    displayType: string,
    mouseHasBeenDown: boolean,
};

const ButtonComponent = React.forwardRef<HTMLButtonElement | null, React.PropsWithChildren<ButtonProps>>(
    ({ children, disabled, onMouseDown, onMouseUp, onClick, onClickCapture, ...props }, ref) => (
        <button
            type="button"
            className={getButtonClasses(props)}
            disabled={disabled}
            onMouseDown={onMouseDown}
            onMouseUp={onMouseUp}
            onClick={onClick}
            onClickCapture={onClickCapture}
            ref={ref}
        >
            <Content>{children}</Content>
        </button>
    )
);

ButtonComponent.defaultProps = {
    disabled: true,
    onClick: () => {},
    onClickCapture: () => {},
    onMouseDown: () => {},
    onMouseUp: () => {},
};

ButtonComponent.displayName = "@fusion/components/general/Button/ButtonComponent";

export default ButtonComponent;
