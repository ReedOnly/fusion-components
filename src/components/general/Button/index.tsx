import * as React from 'react';
import ComponentDisplayContext from '../../contexts/componentDisplayContext';
import ButtonComponent from './components/Button';
import AnchorComponent from './components/Anchor';

type ButtonProps = {
    disabled?: boolean;

    // Variants
    /** Use the contained variant */
    contained?: boolean;
    /** Use the outlined variant */
    outlined?: boolean;
    /** Use the frameless variant */
    frameless?: boolean;

    // Styles
    /** Use the primary coloring */
    primary?: boolean;
    /** Use the signal coloring */
    signal?: boolean;

    // Sizes
    /** Compact button */
    compact?: boolean;
    /** Comfortable button */
    comfortable?: boolean;

    // Link
    /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
    url?: string | null;
    /** Set target="_blank" */
    targetBlank?: boolean;
    onMouseDown?: React.EventHandler<React.SyntheticEvent>;
    onClickCapture?: React.EventHandler<React.SyntheticEvent>;
    onClick?: React.EventHandler<React.SyntheticEvent>;
};

const Button = React.forwardRef<HTMLElement | null, React.PropsWithChildren<ButtonProps>>(
    (props, ref) => {
        // Used to apply "radar" animation on mouse up
        const [mouseHasBeenDown, setMouseHasBeenDown] = React.useState(false);
        const displayType = React.useContext<string>(ComponentDisplayContext);

        const buttonRef =
            (ref as React.RefObject<HTMLElement | null>) || React.createRef<HTMLElement | null>();

        const handleOnMouseDown = (e: React.SyntheticEvent) => {
            const { onMouseDown } = props;

            setMouseHasBeenDown(false);

            if (onMouseDown) {
                onMouseDown(e);
            }
        };

        const handleOnMouseUp = () => {
            setMouseHasBeenDown(true);
            // Focus is automatically set to the button on click,
            // but we don't want it to have focus after the click
            // action is performed. Only when tabbing to the button
            if (buttonRef.current !== null) {
                buttonRef.current.blur();
            }
        };

        return props.url ? (
            <AnchorComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLAnchorElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
                url={props.url}
            />
        ) : (
            <ButtonComponent
                {...props}
                displayType={displayType}
                ref={buttonRef as React.RefObject<HTMLButtonElement | null>}
                mouseHasBeenDown={mouseHasBeenDown}
                onMouseDown={handleOnMouseDown}
                onMouseUp={handleOnMouseUp}
                onClick={props.onClick}
                onClickCapture={props.onClickCapture}
            />
        );
    }
);

Button.displayName = 'Button';

Button.defaultProps = {
    disabled: false,
    contained: false,
    outlined: false,
    frameless: false,
    primary: true,
    signal: false,
    compact: false,
    comfortable: true,
    url: null,
    targetBlank: false,
    onMouseDown: () => {},
    onClickCapture: () => {},
    onClick: () => {},
};

export default Button;
