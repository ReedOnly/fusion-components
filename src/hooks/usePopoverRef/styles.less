.container {
    position: absolute;
    pointer-events: none !important;
}