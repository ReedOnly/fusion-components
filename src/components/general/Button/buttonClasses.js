import classNames from "classnames";
import { componentDisplayTypes } from "../../contexts/ComponentDisplayContext";
import styles from "./styles/index.less";

const getLayoutClasses = ({ block, flex }) => ({
    [styles.block]: block,
    [styles.flex]: flex,
});

const getMouseDownClasses = ({ mouseHasBeenDown }) => ({
    [styles.mouseHasBeenDown]: mouseHasBeenDown,
});

const getButtonVariantClasses = ({ contained, outlined, frameless }) => ({
    [styles.contained]: contained || (!outlined && !frameless), // Default to contained
    [styles.outlined]: outlined,
    [styles.frameless]: frameless,
    [styles.icon]: false,
});

const getButtonStyleClasses = ({ primary, signal }) => ({
    [styles.primary]: primary,
    [styles.signal]: signal,
});

const getButtonSizeClasses = ({ compact, comfortable, displayType }) => ({
    [styles.compact]: displayType === componentDisplayTypes.compact || compact,
    [styles.comfortable]:
        (displayType === componentDisplayTypes.comfortable &&
            displayType !== componentDisplayTypes.comfortable) ||
        (comfortable && !compact),
});

export default props =>
    classNames(
        props.className,
        styles.container,
        getLayoutClasses(props),
        getMouseDownClasses(props),
        getButtonVariantClasses(props),
        getButtonStyleClasses(props),
        getButtonSizeClasses(props)
    );
