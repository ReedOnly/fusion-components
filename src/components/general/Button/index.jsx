import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import ComponentDisplayContext, {
    componentDisplayTypes,
} from "../../contexts/ComponentDisplayContext";
import styles from "./styles.less";

class Button extends Component {
    static propTypes = {
        children: PropTypes.node.isRequired,
        /** Disable the button */
        disabled: PropTypes.bool,

        // Variants
        /** Use the contained variant */
        contained: PropTypes.bool,
        /** Use the outlined variant */
        outlined: PropTypes.bool,
        /** Use the frameless variant */
        frameless: PropTypes.bool,

        // Styles
        /** Use the primary coloring */
        primary: PropTypes.bool,
        /** Use the signal coloring */
        signal: PropTypes.bool,

        // Sizes
        /** Small button */
        small: PropTypes.bool,
        /** Medium button */
        medium: PropTypes.bool,
        /** Large button */
        large: PropTypes.bool,

        // Link
        /** Provide an url and the button wil use <a>-tag instead of <button>-tag */
        url: PropTypes.string,
        /** Set target="_blank" */
        targetBlank: PropTypes.bool,
        onMouseDown: PropTypes.func,
        onClickCapture: PropTypes.func,
        onClick: PropTypes.func,
    };

    static defaultProps = {
        disabled: false,
        contained: false,
        outlined: false,
        frameless: false,
        primary: true,
        signal: false,
        small: true,
        medium: true,
        large: false,
        url: null,
        targetBlank: false,
        onMouseDown: () => {},
        onClickCapture: () => {},
        onClick: () => {},
    };

    state = {
        isMouseDown: false,
        mouseHasBeenDown: false,
    };

    onMouseUp() {
        this.setState({
            isMouseDown: false,
            mouseHasBeenDown: true,
        });

        if (this.button) {
            this.button.blur();
        }
    }

    onMouseDown(e) {
        const { onMouseDown } = this.props;
        if (onMouseDown) {
            onMouseDown(e);
        }

        this.setState({
            isMouseDown: true,
            mouseHasBeenDown: false,
        });
    }

    getMouseDownClasses = () => {
        const { isMouseDown, mouseHasBeenDown } = this.state;

        return {
            [styles.mouseIsDown]: isMouseDown,
            [styles.mouseHasBeenDown]: mouseHasBeenDown,
        };
    };

    getButtonVariantClasses = () => {
        const { contained, outlined, frameless } = this.props;

        return {
            [styles.contained]: contained || (!outlined && !frameless), // Default to contained
            [styles.outlined]: outlined,
            [styles.frameless]: frameless,
            [styles.icon]: false,
        };
    };

    getButtonStyleClasses = () => {
        const { primary, signal } = this.props;
        return {
            [styles.primary]: primary,
            [styles.signal]: signal,
        };
    };

    getButtonSizeClasses = displayType => {
        const { small, medium, large } = this.props;

        return {
            [styles.small]:
                small || displayType === componentDisplayTypes.compact,
            [styles.medium]:
                (medium || (!small && !large)) &&
                displayType !== componentDisplayTypes.compact, // Default to medium
            [styles.large]:
                large && displayType !== componentDisplayTypes.compact,
        };
    };

    getButtonClasses = (displayType, props) =>
        classNames(
            props.className,
            styles.container,
            {
                [styles.block]: props.block,
                [styles.flex]: props.flex,
            },
            this.getMouseDownClasses(props),
            this.getButtonVariantClasses(props),
            this.getButtonStyleClasses(props),
            this.getButtonSizeClasses(displayType)
        );

    renderButtonContent() {
        const { children } = this.props;
        return <span className={styles.button}>{children}</span>;
    }

    renderButton(displayType) {
        const {
            disabled,
            onClick,
            onClickCapture,
            url,
            targetBlank,
        } = this.props;

        const containerClassNames = this.getButtonClasses(displayType, {
            ...this.props,
            ...this.state,
        });

        if (url) {
            return (
                <a
                    className={containerClassNames}
                    disabled={disabled}
                    href={url}
                    target={targetBlank ? "_blank" : "_self"}
                    onMouseDown={() => this.onMouseDown()}
                    onMouseUp={() => this.onMouseUp()}
                    onClickCapture={onClickCapture}
                    ref={button => {
                        this.button = button;
                    }}
                >
                    {this.renderButtonContent()}
                </a>
            );
        }
        return (
            <button
                type="button"
                className={containerClassNames}
                disabled={disabled}
                onClick={onClick}
                onMouseDown={e => this.onMouseDown(e)}
                onMouseUp={() => this.onMouseUp()}
                onClickCapture={onClickCapture}
                ref={button => {
                    this.button = button;
                }}
            >
                {this.renderButtonContent()}
            </button>
        );
    }

    render() {
        return (
            <ComponentDisplayContext.Consumer>
                {displayType => this.renderButton(displayType)}
            </ComponentDisplayContext.Consumer>
        );
    }
}

export default Button;
