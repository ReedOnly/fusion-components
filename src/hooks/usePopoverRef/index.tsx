import * as React from 'react';
import useClickToggleController from '../useClickToggleController';
import useOverlayPortal from '../useOverlayPortal';
import useRelativePositioning from '../useRelativePositioning';
import PopoverContent from '../../components/general/Popover/components/Content';
import * as styles from './styles.less';
import useClickOutsideOverlayPortal from '../useClickOutsideOverlayPortal';

export default <T extends HTMLElement>(
    content: React.ReactNode,
    props?: any
): React.MutableRefObject<T | null> => {
    const popoverContentRef = React.useRef<HTMLDivElement | null>(null);
    const [isOpen, ref, setIsOpen] = useClickToggleController<T>();
    const rect = useRelativePositioning(ref);

    const close = React.useCallback(() => isOpen && setIsOpen(false), [isOpen]);

    useClickOutsideOverlayPortal(close, popoverContentRef.current);

    useOverlayPortal(
        isOpen,
        <div
            ref={popoverContentRef}
            className={styles.container}
            style={{
                width: rect.width,
                height: rect.height,
                top: rect.top,
                left: rect.left,
            }}
        >
            <PopoverContent {...props}>{content}</PopoverContent>
        </div>
    );

    return ref;
};
