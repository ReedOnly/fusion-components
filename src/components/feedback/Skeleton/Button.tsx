import React from 'react';
import classNames from 'classnames';
import styles from './styles.less';
import { useComponentDisplayClassNames } from '@equinor/fusion';
import { ButtonProps } from 'components/general/Button';

type SkeletonButtonProps = ButtonProps & {
    width: number | string;
};

const SkeletonButton: React.FC<SkeletonButtonProps> = ({ width, ...buttonProps }) => {
    const displayClassNames = useComponentDisplayClassNames(styles);
    const barClassNames = classNames(styles.skeleton, displayClassNames, styles.button);

    return <div className={barClassNames} style={{ width }} />;
};

export default SkeletonButton;
