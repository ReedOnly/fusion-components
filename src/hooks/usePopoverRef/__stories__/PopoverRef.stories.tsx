import * as React from 'react';
import { storiesOf } from '@storybook/react';
import Button from '../../../components/general/Button';
import usePopoverRef from '../index';

const PopoverContent = () => <div>This is a popover!</div>;

const PopoverStory = () => {
    const popoverRef = usePopoverRef(<PopoverContent />, {
        // horizontalPosition: "right",
        // verticalPosition: "top",
        justify: 'middle', // start = "left" | middle = "center" | end = "right"
        align: 'end', // start = "top" | middle = "center" | end = "bottom"
        alignArrow: 'start', // start = "left" | middle = "center" | end = "right"
    });

    return (
        <Button ref={popoverRef} primary contained>
            Click for popover
        </Button>
    );
};

storiesOf('Hooks|Popover', module).add('Popover', () => {
    return <PopoverStory />;
});
