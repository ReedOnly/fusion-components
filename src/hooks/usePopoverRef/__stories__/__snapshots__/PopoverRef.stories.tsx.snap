// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hooks|Popover Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <button
      className="container contained primary comfortable"
      disabled={false}
      onClick={[Function]}
      onClickCapture={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      type="button"
    >
      <span
        className="button"
      >
        Click for popover
      </span>
    </button>
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
